### PR TITLE
Implement %elif

### PR DESCRIPTION
**Pull request by pavlinamv**

### CODE REVIEW

File: build/parseSpec.c [hunk 1]
```diff
@@ -501,17 +506,24 @@ int readLine(rpmSpec spec, int strip)
 	goto retry;
     }
 
-after_classification:
-    if (match != -1) {
+    if (lineType->id & (LINE_IF | LINE_IFARCH | LINE_IFNARCH | LINE_IFOS |
+	    LINE_IFNOS)) {
```
Comment (pmatilai) [inline build/parseSpec.c]: This calls for a bitmask, something like LINE_IFANY, which can be used to trap all the types with one symbol. Makes the line shorter and more readable. (fwiw no I'm not reviewing, just taking a look)

File: build/parseSpec.c [hunk 2]
```diff
@@ -210,8 +210,12 @@ static int expandMacrosInSpecBuf(rpmSpec spec, int strip)
...
     }
 
+    /* Don't expand macros after %elif (resp. %elifarch, %elifos) in a false branch */
+    if (condition && (condition->id & (LINE_ELIF | LINE_ELIFOS | LINE_ELIFARCH))) {
+	if (!spec->readStack->readable)
+	    return 0;
     /* Don't expand macros (eg. %define) in false branch of %if clause */
-    if (!spec->readStack->reading)
+    } else if (!spec->readStack->reading)
 	return 0;
```
Comment (pmatilai) [inline build/parseSpec.c]: If one branch of if uses { } blocks, all the branches should have them.

File: build/parseSpec.c [hunk 3]
```diff
@@ -447,31 +451,32 @@ int readLine(rpmSpec spec, int strip)
...
-	if (match < 0) {
-	    rpmlog(RPMLOG_ERR,
-			_("%s:%d: bad %%if condition\n"),
-			ofi->fileName, ofi->lineNum);
-	    return PART_ERROR;
+    } else if (lineType->id & (LINE_IF | LINE_ELIF)) {
+	s += lineType->textLen;
+	if (((lineType->id == LINE_IF) && (spec->readStack->reading)) ||
+	    ((lineType->id == LINE_ELIF) && (spec->readStack->readable))) {
+	    match = parseExpressionBoolean(s);
```
Comment (pmatilai) [inline build/parseSpec.c]: match = .. is at the same indentation level as the continued condition, please fix. Again, helper variable or two can be used to shorten the condition and will likely make it nicer overall.

**Comment by pavlinamv** [comment]
Panu's comments are incorporated.

File: build/parseSpec.c [hunk 4]
```diff
@@ -501,17 +511,23 @@ int readLine(rpmSpec spec, int strip)
...
+    if (lineType->id & (LINE_IFANY)) {
 	rl = xmalloc(sizeof(*rl));
 	rl->reading = spec->readStack->reading && match;
 	rl->next = spec->readStack;
 	rl->lineNum = ofi->lineNum;
+	rl->readable = (!rl->reading) && (spec->readStack->reading);
 	rl->lastConditional = lineType;
 	spec->readStack = rl;
 	spec->line[0] = '\0';
+    } else if (lineType->id & (LINE_ELIF | LINE_ELIFARCH | LINE_ELIFOS)) {
```
Comment (pmatilai) [inline build/parseSpec.c]: The various ELIF's are another candidate for a bit mask, this set is repeated at least twice in the patch (used earlier in the "dont expand in false branch" check). In general, when addressing something, it pays to look for similar cases/opportunities/bugs, they're quite often there.

Comment (pavlinamv) [reply build/parseSpec.c]: It is changed in the new version.

**Comment by pavlinamv** [comment]
Corrected the commit message of the first commit.

**Comment by ffesti** [comment]
Ok, Rebased and merged.

**Pull request closed without merging**